### DISCUSSION TITLE
(Positional) argument transformation

### DISCUSSION THREAD

**Discussion by abadrinath15** [microsoft/pyright#7125] [Q&A] [answered]
I'm trying to add some typing to an existing code base that definitely does some... interesting stuff with decorators. Here's an example of a behavior I'm struggling to correctly type an offender: it _augments_ the existing types of all arguments. A simple example would be `T` -> `Optional[T]`. 

Concretely, here's an example:
```python

# don't know what to do here
def can_be_none(f): 
    ...

@can_be_null
def some_func(a: int, b: str) -> int:
    ...

some_func(2, 'a') # should typecheck ok
some_func(None, 'a') # also should typecheck ok
```

Is it even possible to express this? I don't think ParamSpec or TypevarTuple support transformations, and Concatenate only seems to support a deterministic number of arguments?

Thank you!

**Comment by erictraut** [comment, marked as answer]
No, this isn't possible to express in today's type system, at least not in a generalized way. If you have a limited set of signatures that you want to support in your decorator, you could define an overload for the decorator function.

**Reply by abadrinath15** [reply]
Got it. Thanks!
